iceoryx_posh/source/roudi/roudi_process.cpp

```diff
@@ -126,21 +134,151 @@ ProcessManager::ProcessManager(RouDiMemoryInterface& roudiMemoryInterface,
     m_memoryManagerOfCurrentProcess = m_segmentInfo.m_memoryManager;
 }
 
-void ProcessManager::killAllProcesses() noexcept
+void ProcessManager::killAllProcesses(const units::Duration processKillDelay) noexcept
 {
-    std::lock_guard<std::mutex> g(m_mutex);
+    std::lock_guard<std::mutex> lockGuard(m_mutex);
+    cxx::vector<bool, MAX_PROCESS_NUMBER> processStillRunning(m_processList.size(), true);
+    uint64_t i{0};
+    bool haveAllProcessesFinished{false};
+    posix::Timer finalKillTimer(processKillDelay);
 
-    // send SIGTERM to all running applications
-    typename ProcessList_t::iterator it = m_processList.begin();
-    const typename ProcessList_t::iterator itEnd = m_processList.end();
+    auto awaitProcessTermination = [&]() {
+        bool shouldCheckProcessState = true;
+        finalKillTimer.resetCreationTime();
 
-    for (; itEnd != it; ++it)
+        while (!haveAllProcessesFinished && !finalKillTimer.hasExpiredComparedToCreationTime())
```

Comment:
It's still a bit hard for me to quickly grasp what's happening here again after forgetting... Maybe I'm just dumb.
Could we add some short comments or self-documenting functions for dumb people like me? e.g. What is this while loop doing ? Waiting for the applications to shut themselves down ?

Reply:
That's true.

Reply:
So... this was resolved but still no comment in code ?

Reply:
I added a comment just before `while`, that's why it's resolved. Check out second to last commit 1e130ed.